File: GFG/Asteroid Collision/Asteroid Collision.java

class Solution {
    public static int[] asteroidCollision(int N, int[] arr) {
        // code here
        Stack<Integer> st=new Stack<>();
        for(int num:arr){
            if(num<0){
                while(!st.isEmpty() && st.peek()>0 && st.peek()<Math.abs(num)){
                    st.pop();
                }
                
                if(!st.isEmpty()){
                    if(st.peek()==Math.abs(num)){
                        st.pop();
                    }else if(st.peek()<0){
                        st.push(num);
                    }
                }else if(st.isEmpty()){
                    st.push(num);
                }
            }else{
                st.push(num);
            }
        }
        int[] ans =new int[st.size()];
        int i=st.size()-1;
        while(st.size()>0)ans[i--]=st.pop();
        return ans;
    }
}

File: GFG/Asteroid Collision/Asteroid Collision.cpp
class Solution {
  public:
    vector<int> asteroidCollision(int N, vector<int> &asteroids) {
        stack<int> s;
        int i = 0;
        while(i < N){
            if(s.empty()){
                s.push(asteroids[i]);
                ++i;
            }
            else{
                if((asteroids[i] > 0 && s.top() > 0) || 
                (asteroids[i] < 0 && s.top() < 0) ||
                (asteroids[i] > 0 && s.top() < 0)){
                    s.push(asteroids[i]);
                    ++i;
                }
                else if(asteroids[i] < 0 && s.top() > 0){
                    int a = s.top();
                    s.pop();
                    
                    if(a == abs(asteroids[i]))
                        ++i;
                    else if(a > abs(asteroids[i])){
                        s.push(a);
                        ++i;
                    }
                    else if(a < abs(asteroids[i])){
                        continue;
                    }
                }
            }
        }
        vector<int> ans;
        while(!s.empty()){
            int a = s.top();
            s.pop();
            ans.push_back(a);
        }
        
        reverse(ans.begin(), ans.end());
        
        return ans;
    }
};
